### ISSUE TITLE
Use distance functions with native pytorch triplet loss

### ISSUE THREAD

**Issue by Pharaun85** [KevinMusgrave/pytorch-metric-learning#241]
There is any way to use some of the package distance functions, like SNR in the Pytorch loss function TripletMarginWithDistanceLoss? I did not found in the documentation and it did not work in my code. I know it can be used with the specific TripletMarginLoss of this package but, in my code, the triplets come from different sources and I don't know how to replicate that with TripletMarginLoss function.
Being more specific my data loader provides three images, one for the anchor (usually an RGB image) and two for the positive and negative items, always a grayscale image with a shape. This type of triplets is the ones that I use to calculate the loss, and as far as I know, this is not possible with the package TripletMarginLoss. Am I wrong?

I'm using it like this:
```
criterion = torch.nn.TripletMarginWithDistanceLoss(distance_function=SNRDistance(),
                                                                   margin=args.margin, reduction='mean')
```

Thanks

**Comment by KevinMusgrave** [comment]
Since your dataloader already returns all the triplets that you want to use, this is probably the best solution:

```python
for (a, p, n) in dataloader:
    # create anchor, positive, negative indices
    batch_size = len(a)
    idx = torch.arange(batch_size*3)
    a_idx = idx[ : batch_size]
    p_idx = idx[batch_size : batch_size*2]
    n_idx = idx[batch_size*2 : batch_size*3]

    # get embeddings from model
    # ... 
    # put all embeddings into one tensor
    embeddings = torch.cat([a_emb, p_emb, n_emb], axis=0)
    labels = torch.arange(batch_size)
    labels = torch.cat([labels, labels, labels], axis=0)
    
    # the loss function will use only the triplets formed by (a_idx, p_idx, n_idx)
    loss = loss_fn(embeddings, labels, (a_idx, p_idx, n_idx))
```

If your data comes from 2 sources but your dataloader doesn't return triplets, you can use ```get_all_triplets_indices```:
```python
from pytorch_metric_learning.utils import loss_and_miner_utils as lmu
# anchors will come from labels
# positives and negatives will come from ref_labels
a_idx, p_idx, n_idx = lmu.get_all_triplets_indices(labels, ref_labels)
```

If you want to use SNR with the PyTorch loss function ```TripletMarginWithDistanceLoss```, I think this should work:

```python
def SNR(x, y):
    return torch.var(x - y) / torch.var(x)

criterion = torch.nn.TripletMarginWithDistanceLoss(distance_function=SNR, margin=args.margin, reduction='mean')
```

**Comment by Pharaun85** [comment]
Thanks for your quick response! 
Since my dataloader return samples like this:
```
sample = {'anchor': rgb_image,
                  'positive': OSM_positive[0],
                  'negative': OSM_negative[0],
                  'label_anchor': label_positive,
                  'label_positive': label_positive,
                  'label_negative': label_negative,
                  }
```
I think it would be easy to implement your first solution and thus be able to use more distance functions and even miners reducers and regularizers.
